pkg/workspace/assert.go

```diff
@@ -0,0 +1,167 @@
+// Checks, validates, and runs the assertions, any violations from running the assertions are collected and stored.
+//
+// This visitor returns and error if any assertion is not well-formed,
+// otherwise, returns nil.
+func (a *assertChecker) Visit(node yamlmeta.Node) error {
+	nodeAnnotations := template.NewAnnotations(node)
+	if nodeAnnotations.Has(AnnotationAssertValidate) {
+		validations, syntaxErr := newValidationsFromAssertValidateAnnotation(nodeAnnotations[AnnotationAssertValidate], node, a.thread)
+		if syntaxErr != nil {
+			return syntaxErr
+		}
+		for _, v := range validations {
+			err := v.Validate(node)
+			if err != nil {
+				a.violations = append(a.violations, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func newValidationsFromAssertValidateAnnotation(annotation template.NodeAnnotation, n yamlmeta.Node, thread *starlark.Thread) ([]Validation, error) {
+	var validations []Validation
+	validationPosition := createAssertAnnotationPosition(annotation.Position, n)
+
+	if len(annotation.Kwargs) != 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found keyword argument (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	if len(annotation.Args) == 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found no arguments (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	for _, arg := range annotation.Args {
+		validationTuple, ok := arg.(starlark.Tuple)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found: %v (by %v)", AnnotationAssertValidate, arg.String(), validationPosition.AsCompactString())
+		}
+		if len(validationTuple) != 2 {
+			return nil, fmt.Errorf("Expected @%v 2-tuple, but found tuple with length %v (by %v)", AnnotationAssertValidate, len(validationTuple), validationPosition.AsCompactString())
+		}
+		message, ok := validationTuple[0].(starlark.String)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have descriptive string as the first item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[0].Type(), validationPosition.AsCompactString())
+		}
+		lambda, ok := validationTuple[1].(starlark.Callable)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have a validation function as the second item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[1].Type(), validationPosition.AsCompactString())
+		}
+		validations = append(validations, Validation{
+			thread:   thread,
+			msg:      message.String(),
+			f:        lambda,
+			position: validationPosition,
+		})
+	}
+
+	return validations, nil
+}
+
+// Validation is an assertion from an @assert/validate annotation.
+// One @assert/validate annotation can have multiple Validations.
+type Validation struct {
+	thread   *starlark.Thread
```

Comment:
The insights that I gained from looking at other thread creations in the code suggested that the thread be created higher up in the code path (currently created in ProcessAndRunAssertions). The thread is currently injected all the way down to a Validation, however I also experimented with it as a parameter to the Validate function: `Validation.Validate(Node, starlark.Thread)`. The thread would then be supplied to the validate function via the `AssertChecker`. Thoughts?

Reply:
Prior art suggests to me that a new thread is needed for each evaluation:
https://github.com/vmware-tanzu/carvel-ytt/blob/b50fd584133ce57a147c9fa050bd4acea684a259/pkg/workspace/template_loader.go#L197

So, from what I can tell, the more relevant part of Thread is its ability to carry context along the flow-of-execution (via `SetLocal()`/`Local()`).

Would we ever want to _share_ context in this way across assertion calls (or from a higher level process to assertion calls...)?
Conversely, do we want to specifically _prevent_ sharing context in this way across assertion calls?

🤔

Reply:
Since I cannot produce a reason to either explicitly include or exclude context from the TemplateLoader, I think it would be best to include the Loader when creating a new thread, so that we have the option later to include or not include the context that the Loader contains.